File: dotnet/src/Envdoctor/Models/EnvironmentFile.cs
namespace Envdoctor.Models;

public enum FileFormat
{
    Dotenv,
    DockerCompose,
    GithubActions,
    Kubernetes,
    Source,
}

/// The parsed contents of a single file, normalized to envdoctor's model.
public sealed class EnvironmentFile
{
    public required string FilePath { get; set; }
    public FileFormat Format { get; set; }
    public string? Environment { get; set; }
    public List<EnvironmentVariable> Variables { get; set; } = new();
    public List<EnvironmentVariable> Usages { get; set; } = new();

    public EnvironmentFile Clone() =>
        new()
        {
            FilePath = FilePath,
            Format = Format,
            Environment = Environment,
            Variables = Variables.Select(v => v.Clone()).ToList(),
            Usages = Usages.Select(v => v.Clone()).ToList(),
        };

    /// Names defined in a file, deduplicated, in first-seen order.
    public List<string> DefinedNames()
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var v in Variables)
        {
            if (seen.Add(v.Name))
                result.Add(v.Name);
        }
        return result;
    }

    /// Names used in a file, deduplicated, in first-seen order.
    public List<string> UsedNames()
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var v in Usages)
        {
            if (seen.Add(v.Name))
                result.Add(v.Name);
        }
        return result;
    }
}

File: swift/Sources/EnvdoctorKit/Models/EnvironmentFile.swift
import Foundation

public enum FileFormat: String {
    case dotenv, dockerCompose = "docker-compose", githubActions = "github-actions", kubernetes, source
}

/// The parsed contents of a single file, normalized to envdoctor's model.
public struct EnvironmentFile {
    /// Path the file was read from.
    public var filePath: String
    public var format: FileFormat
    /// Environment label for dotenv files ("development", "production", ...).
    public var environment: String?
    public var variables: [EnvironmentVariable]
    public var usages: [EnvironmentVariable]

    public init(filePath: String, format: FileFormat, environment: String? = nil, variables: [EnvironmentVariable], usages: [EnvironmentVariable]) {
        self.filePath = filePath
        self.format = format
        self.environment = environment
        self.variables = variables
        self.usages = usages
    }
}

/// Names defined in a file, deduplicated, in first-seen order.
public func definedNames(_ file: EnvironmentFile) -> [String] {
    var seen = Set<String>()
    var out: [String] = []
    for v in file.variables where !seen.contains(v.name) {
        seen.insert(v.name)
        out.append(v.name)
    }
    return out
}

/// Names used in a file, deduplicated, in first-seen order.
public func usedNames(_ file: EnvironmentFile) -> [String] {
    var seen = Set<String>()
    var out: [String] = []
    for v in file.usages where !seen.contains(v.name) {
        seen.insert(v.name)
        out.append(v.name)
    }
    return out
}
